### ISSUE TITLE
ERROR: Failed building wheel for mpi4py

### ISSUE THREAD

**Issue by gauravkuppa** [christopher-hsu/scalableMARL#2]
Thank you for your paper and your work. 

I am trying to setup this code. I am running a Docker container with Python 3.8 and Ubuntu 20.04. When I try to run `pip3 install --trusted-host pypi.python.org -r requirements.txt`, I get an error saying that the OpenCV version does not exist. I change it to opencv=4.1.2.30. When I try to run the command again, I get faced with the error message shown below. When I try to run `sudo apt install libopenmpi-dev` as recommended [here](https://stackoverflow.com/questions/28440834/error-when-installing-mpi4py), I get another compounded error.

I would appreciate any help to resolve this issue.
```
ERROR: Command errored out with exit status 1:
command: /usr/bin/python3.8 /tmp/tmpbe2eo2l7 build_wheel /tmp/tmpfnvwy74h
     cwd: /tmp/pip-install-3fm473kh/mpi4py
Complete output (148 lines):
running bdist_wheel
running build
running build_src
running build_py
creating build
creating build/lib.linux-x86_64-3.8
creating build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/bench.py -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/__main__.py -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/run.py -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/__init__.py -> build/lib.linux-x86_64-3.8/mpi4py
creating build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/_core.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/__main__.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/_base.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/_lib.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/aplus.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/__init__.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/server.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/pool.py -> build/lib.linux-x86_64-3.8/mpi4py/futures
creating build/lib.linux-x86_64-3.8/mpi4py/util
copying src/mpi4py/util/pkl5.py -> build/lib.linux-x86_64-3.8/mpi4py/util
copying src/mpi4py/util/dtlib.py -> build/lib.linux-x86_64-3.8/mpi4py/util
copying src/mpi4py/util/__init__.py -> build/lib.linux-x86_64-3.8/mpi4py/util
copying src/mpi4py/py.typed -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/__main__.pyi -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/MPI.pyi -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/dl.pyi -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/run.pyi -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/bench.pyi -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/__init__.pyi -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/MPI.pxd -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/__init__.pxd -> build/lib.linux-x86_64-3.8/mpi4py
copying src/mpi4py/libmpi.pxd -> build/lib.linux-x86_64-3.8/mpi4py
creating build/lib.linux-x86_64-3.8/mpi4py/include
creating build/lib.linux-x86_64-3.8/mpi4py/include/mpi4py
copying src/mpi4py/include/mpi4py/mpi4py.MPI_api.h -> build/lib.linux-x86_64-3.8/mpi4py/include/mpi4py
copying src/mpi4py/include/mpi4py/mpi4py.h -> build/lib.linux-x86_64-3.8/mpi4py/include/mpi4py
copying src/mpi4py/include/mpi4py/mpi4py.MPI.h -> build/lib.linux-x86_64-3.8/mpi4py/include/mpi4py
copying src/mpi4py/include/mpi4py/mpi4py.i -> build/lib.linux-x86_64-3.8/mpi4py/include/mpi4py
copying src/mpi4py/include/mpi4py/mpi.pxi -> build/lib.linux-x86_64-3.8/mpi4py/include/mpi4py
copying src/mpi4py/futures/pool.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/aplus.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/_lib.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/__main__.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/server.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/_core.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/futures/__init__.pyi -> build/lib.linux-x86_64-3.8/mpi4py/futures
copying src/mpi4py/util/pkl5.pyi -> build/lib.linux-x86_64-3.8/mpi4py/util
copying src/mpi4py/util/__init__.pyi -> build/lib.linux-x86_64-3.8/mpi4py/util
copying src/mpi4py/util/dtlib.pyi -> build/lib.linux-x86_64-3.8/mpi4py/util
running build_clib
MPI configuration: [mpi] from 'mpi.cfg'
checking for library 'lmpe' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -llmpe -o _configtest
/usr/bin/ld: cannot find -llmpe
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
building 'mpe' dylib library
creating build/temp.linux-x86_64-3.8
creating build/temp.linux-x86_64-3.8/src
creating build/temp.linux-x86_64-3.8/src/lib-pmpi
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c src/lib-pmpi/mpe.c -o build/temp.linux-x86_64-3.8/src/lib-pmpi/mpe.o
creating build/lib.linux-x86_64-3.8/mpi4py/lib-pmpi
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,--no-as-needed build/temp.linux-x86_64-3.8/src/lib-pmpi/mpe.o -o build/lib.linux-x86_64-3.8/mpi4py/lib-pmpi/libmpe.so
checking for library 'vt-mpi' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -lvt-mpi -o _configtest
/usr/bin/ld: cannot find -lvt-mpi
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
checking for library 'vt.mpi' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -lvt.mpi -o _configtest
/usr/bin/ld: cannot find -lvt.mpi
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
building 'vt' dylib library
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c src/lib-pmpi/vt.c -o build/temp.linux-x86_64-3.8/src/lib-pmpi/vt.o
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,--no-as-needed build/temp.linux-x86_64-3.8/src/lib-pmpi/vt.o -o build/lib.linux-x86_64-3.8/mpi4py/lib-pmpi/libvt.so
checking for library 'vt-mpi' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -lvt-mpi -o _configtest
/usr/bin/ld: cannot find -lvt-mpi
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
checking for library 'vt.mpi' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -lvt.mpi -o _configtest
/usr/bin/ld: cannot find -lvt.mpi
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
building 'vt-mpi' dylib library
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c src/lib-pmpi/vt-mpi.c -o build/temp.linux-x86_64-3.8/src/lib-pmpi/vt-mpi.o
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,--no-as-needed build/temp.linux-x86_64-3.8/src/lib-pmpi/vt-mpi.o -o build/lib.linux-x86_64-3.8/mpi4py/lib-pmpi/libvt-mpi.so
checking for library 'vt-hyb' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -lvt-hyb -o _configtest
/usr/bin/ld: cannot find -lvt-hyb
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
checking for library 'vt.ompi' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -lvt.ompi -o _configtest
/usr/bin/ld: cannot find -lvt.ompi
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
building 'vt-hyb' dylib library
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -c src/lib-pmpi/vt-hyb.c -o build/temp.linux-x86_64-3.8/src/lib-pmpi/vt-hyb.o
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,--no-as-needed build/temp.linux-x86_64-3.8/src/lib-pmpi/vt-hyb.o -o build/lib.linux-x86_64-3.8/mpi4py/lib-pmpi/libvt-hyb.so
running build_ext
MPI configuration: [mpi] from 'mpi.cfg'
checking for dlopen() availability ...
checking for header 'dlfcn.h' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/include/python3.8 -c _configtest.c -o _configtest.o
success!
removing: _configtest.c _configtest.o
success!
checking for library 'dl' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/include/python3.8 -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -Lbuild/temp.linux-x86_64-3.8 -ldl -o _configtest
success!
removing: _configtest.c _configtest.o _configtest
checking for function 'dlopen' ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/include/python3.8 -c _configtest.c -o _configtest.o
x86_64-linux-gnu-gcc -pthread _configtest.o -Lbuild/temp.linux-x86_64-3.8 -ldl -o _configtest
success!
removing: _configtest.c _configtest.o _configtest
building 'mpi4py.dl' extension
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DHAVE_DLFCN_H=1 -DHAVE_DLOPEN=1 -I/usr/include/python3.8 -c src/dynload.c -o build/temp.linux-x86_64-3.8/src/dynload.o
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-x86_64-3.8/src/dynload.o -Lbuild/temp.linux-x86_64-3.8 -ldl -o build/lib.linux-x86_64-3.8/mpi4py/dl.cpython-38-x86_64-linux-gnu.so
checking for MPI compile and link ...
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/include/python3.8 -c _configtest.c -o _configtest.o
_configtest.c:2:10: fatal error: mpi.h: No such file or directory
    2 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
failure.
removing: _configtest.c _configtest.o
error: Cannot compile MPI programs. Check your configuration!!!
----------------------------------------
ERROR: Failed building wheel for mpi4py
```

**Comment by christopher-hsu** [comment]
Please try again, I saw there was an issue with the Pillow version I was using which might mess up the previous dependencies.

**Comment by gauravkuppa** [comment]
I tried again, and I got past the installation portion. However, `pip install` still causes this error related to mpi4py

```
ERROR: Failed building wheel for mpi4py
  Building wheel for audioread (setup.py) ... done
  Created wheel for audioread: filename=audioread-2.1.9-py3-none-any.whl size=23142 sha256=21cae8fd1e73d1f3236baa384702ad7b0d282cbc1a98a5ae388d23d00e2aa5c7
  Stored in directory: /home/docker/.cache/pip/wheels/49/5a/e4/df590783499a992a88de6c0898991d1167453a3196d0d1eeb7
  Building wheel for future (setup.py) ... done
  Created wheel for future: filename=future-0.18.2-py3-none-any.whl size=491058 sha256=db6a3d7006f7674410b73dd19846caf75d29907ed46c1b5a627ba38ef5729662
  Stored in directory: /home/docker/.cache/pip/wheels/8e/70/28/3d6ccd6e315f65f245da085482a2e1c7d14b90b30f239e2cf4
  Building wheel for gym (pyproject.toml) ... done
  Created wheel for gym: filename=gym-0.23.1-py3-none-any.whl size=701378 sha256=6a9d31a7ba3348a2227f31393e3715bc25036b6feca8df3c210a5ad705109ea7
  Stored in directory: /home/docker/.cache/pip/wheels/78/28/77/b0c74e80a2a4faae0161d5c53bc4f8e436e77aedc79136ee13
Successfully built librosa tabulate filterpy audioread future gym
Failed to build mpi4py
ERROR: Could not build wheels for mpi4py, which is required to install pyproject.toml-based projects
```

**Comment by christopher-hsu** [comment]
do you need to use docker? i haven't tested the interactions of packages with it. also mpi4py has always caused issues and that why i have a lot of version requirements

**Comment by christopher-hsu** [comment]
a suggestion would be to follow the setup for https://github.com/openai/spinningup as my code is built on their style of framework. They introduced mpi4py for logging purposes and might not be needed if you don't care about seeing logging info in your terminal. You can comment all that out and simply use tensorboard which is also used in this project.

**Comment by gauravkuppa** [comment]
I don't think Docker is the source of errors. I tested it on a VM and a local machine, and faced the same errors. Actually, installing `sudo apt-get update && sudo apt-get install libopenmpi-dev` from spinningup did the trick to install mpi4py.

I also found some other errors that were fixed by adding stuff to the requirements.txt. Specifically, I commented out the version of `numba` and `joblib`. I added `h5py` and `typing-extensions`.

Here is the new requirements.txt: [requirements.txt](https://github.com/christopher-hsu/scalableMARL/files/8329454/requirements.txt)


Thanks!
